File: c/brain.c
#include<stdlib.h>
#include<stdio.h>
#include<time.h>
#include<assert.h>
#include"brain.h"
#include"utils.h"

//John Balis 2018
//for support email balisujohn@gmail.com



/*
   copies an existing brain, preserving state;
 */
brain * forkBrain(brain * oldBrain)
{
	brain * b  = malloc(sizeof(brain));
	b->neuronCount = oldBrain->neuronCount;
	b->neurons = malloc(sizeof(neuron) * oldBrain->neuronCount);
	for(int i = 0; i < oldBrain->neuronCount; i++)
	{


		b->neurons[i].fired = oldBrain->neurons[i].fired;
		b->neurons[i].activationPotential =  oldBrain->neurons[i].activationPotential;
		b->neurons[i].excitation = oldBrain->neurons[i].excitation;
		b->neurons[i].targetCount = oldBrain->neurons[i].targetCount;
		b->neurons[i].age=oldBrain->neurons[i].age;
		b->neurons[i].activationDuration=oldBrain->neurons[i].activationDuration;
		b->neurons[i].mostRecentActivation= oldBrain->neurons[i].mostRecentActivation;
		b->neurons[i].targets = malloc(sizeof(int)*( b->neurons[i].targetCount));
		for (int c = 0; c < b->neurons[i].targetCount;c++)
		{
			b->neurons[i].targets[c] = oldBrain->neurons[i].targets[c];

		}
		b->neurons[i].potentialWeights = malloc(sizeof(float) * (b->neurons[i].targetCount));
		b->neurons[i].potentialTimes = malloc(sizeof(float) * (b->neurons[i].targetCount));
		for (int c = 0; c < b->neurons[i].targetCount; c ++)
		{
			b->neurons[i].potentialWeights[c] = oldBrain->neurons[i].potentialWeights[c];
			b->neurons[i].potentialTimes[c] = oldBrain->neurons[i].potentialTimes[c] ;
		}

	}
	return b; 

}

/*
frees a neuron

*/

void freeNeuron(neuron *  n)
{
	free(n->potentialWeights);
	free(n->potentialTimes);
	free(n->targets);

}


/*
   frees a brain struct

 */
void freeBrain(brain * b )
{


	for(int i = 0 ; i < b->neuronCount; i++)
	{
		freeNeuron(&(b->neurons[i]));
	}

	free(b->neurons);
	free(b);
}

/*

sets mutation params to comfortable defaults

*/
mutationParams *  initializeDefaultMutationParams()
{

	mutationParams * m = malloc(sizeof(mutationParams));
	m->order = malloc(sizeof(char) * 6);
	m->order[1] = 'c';
	m->order[1] = 't';
	m->order[2] = 'p';
	m->order[0] = 's';
	//m->order[4] = 'h';
	m->order[4] = '\0';
	m->swapProb = 0.1;
	m->neuronCountProb = 0.125;
	m->neuronCountBias = .5;
	m->targetCountProb = 0.25;
	m->targetCountBias = 0.5;
	m->retargetProb = 0.25;
	m->potentialProb = 0.1;
	m->potentialStrength =0.1;
	m->thresholdProb = 0.1;
	m->thresholdStrength = 0.1;
	m->initialNeuronCount = 10;
	m->targetLimit = 5;
	m->minInputCount = 5;
	m->minOutputCount = 5;
	return m;
}


/*
sets general params to comfortable defaults
*/
params * initializeDefaultParams()
{
	params * p = malloc(sizeof(params));
	p->mParams = initializeDefaultMutationParams();
	p->CONTENT_MASK = 0;
	return p;
}


/*
   initializes a random neuron. requires the number of other neurons as
   a range for selecting targets
 */

void initializeNeuron(neuron * n,  int neuronCount/* , int targetCountLimit*/)
{
	n->fired = 0;	
	n->age=0;
	n->activationPotential = randFloat();
	n->excitation = 0.0 ;//randFloat();
	n->activationDuration=0.05;
	n->mostRecentActivation= -100.0;
	n->targetCount = randRange(neuronCount-1)/2;
	n->targets = malloc(sizeof(int)* (n->targetCount));


	for (int c = 0; c < n->targetCount;c++)
	{

		//	do
		//	{ 
		n->targets[c] = randRange(neuronCount-1)+1;
		//	}
		//	while(n->targets[c] == i || ((i ==0 || i == 1) && n->targets[c] ==neuronCount-1)); 


	}



	n->potentialWeights = malloc(sizeof(float) * (neuronCount-1));
	n->potentialTimes = malloc(sizeof(float) * (neuronCount-1));
	for (int c = 0; c < n->targetCount; c ++)
	{
		n->potentialWeights[c] = (randFloat()*2)-1;
		n->potentialTimes[c] = 0.1;
	}



}

/*
   copies the contents of a neuron into another neuron
 */
void copyNeuron(neuron * original, neuron * replica)
{
	replica->age = original->age;
	replica->targetCount = original->targetCount;
	replica->fired = original->fired;
	replica->activationDuration = original->activationDuration;
	replica->activationPotential = original->activationPotential;
	replica->excitation = original->excitation;
	replica->mostRecentActivation= original->mostRecentActivation;
	replica->targets = original->targets;
	replica->potentialWeights= original->potentialWeights;
	replica->potentialTimes=original->potentialTimes;
}







/*
   allocates and initializes a basic random brain with NEURON_COUNT starting neurons
 */
brain * generateBasicBrain()
{
	assert(NEURON_COUNT > 1);


	brain * b  = malloc(sizeof(brain));
	b->neuronCount = NEURON_COUNT;
	b->neurons = malloc(sizeof(neuron) * NEURON_COUNT);
	for(int i = 0; i < NEURON_COUNT; i++)
	{

		initializeNeuron(&(b->neurons[i]), NEURON_COUNT);
	} 
	return b;

}

/*
generates a brain with few connections
*/
brain * generateSparseBrain(mutationParams * m)
{
	assert(m->initialNeuronCount > 1);


	brain * b  = malloc(sizeof(brain));
	b->neuronCount = m->initialNeuronCount;
	b->neurons = malloc(sizeof(neuron) * m->initialNeuronCount );
	for(int i = 0; i < m->initialNeuronCount ; i++)
	{

		initializeNeuron(&(b->neurons[i]), 2);
	} 
	return b;

}



/*
DEPRECATED:  
generates a brain hardcoded to perform Xor.
this doesn't work with a non-linear gate function, so it is deprecated
 */
brain * generateXorBrain()
{
	brain * b  = malloc(sizeof(brain));
	b->neuronCount = 6;
	b->neurons = malloc(sizeof(neuron) * 6);
	for(int i = 0; i < 6; i++)
	{
		b->neurons[i].fired = 0;	
		b->neurons[i].age=0;
		b->neurons[i].activationPotential = 1.0;
		b->neurons[i].excitation = 0.0 ;
		b->neurons[i].activationDuration=0.05;
		b->neurons[i].mostRecentActivation= -100.0;
	} 



	neuron * neurons = b->neurons;


	neurons[0].targetCount = 0;


	neurons[1].targetCount = 2;
	neurons[1].targets = malloc(sizeof(int) *  neurons[1].targetCount);
	neurons[1].potentialWeights = malloc(sizeof(float) * neurons[1].targetCount);
	neurons[1].potentialTimes = malloc(sizeof(float) * neurons[1].targetCount); 
	neurons[1].targets[0] = 4;
	neurons[1].targets[1] = 5;
	neurons[1].potentialWeights[0] =  1.0;
	neurons[1].potentialWeights[1] = -2.0;


	neurons[2].targetCount = 2;
	neurons[2].targets = malloc(sizeof(int) *  neurons[2].targetCount);
	neurons[2].potentialWeights = malloc(sizeof(float) * neurons[2].targetCount);
	neurons[2].potentialTimes = malloc(sizeof(float) * neurons[2].targetCount); 
	neurons[2].targets[0] = 4;
	neurons[2].targets[1] = 5;
	neurons[2].potentialWeights[0] =  -2.0;
	neurons[2].potentialWeights[1] = 1.0;


	neurons[3].targetCount = 0;


	neurons[4].targetCount = 1;
	neurons[4].targets = malloc(sizeof(int) *  neurons[4].targetCount);
	neurons[4].potentialWeights = malloc(sizeof(float) * neurons[4].targetCount);
	neurons[4].potentialTimes = malloc(sizeof(float) * neurons[4].targetCount); 
	neurons[4].targets[0] = 3;
	neurons[4].potentialWeights[0] =  1.0;


	neurons[5].targetCount = 1;
	neurons[5].targets = malloc(sizeof(int) *  neurons[5].targetCount);
	neurons[5].potentialWeights = malloc(sizeof(float) * neurons[5].targetCount);
	neurons[5].potentialTimes = malloc(sizeof(float) * neurons[5].targetCount); 
	neurons[5].targets[0] = 3;
	neurons[5].potentialWeights[0] =  1.0;





	return b;

}


//function to mutate neuron count, probability encodes the probability of occurance, and the bias encodes the 
//probability of removing v.s. adding a neuron. 0 guarantees removal, 1 guarantees addition.

void neuronCountMutation(brain * b, int minInputCount, int minOutputCount, float probability, float bias)
{
	if(randFloat() < probability )
	{
		if(randFloat() < bias)
		{
			b->neuronCount++;
			b->neurons = realloc(b->neurons, sizeof(neuron) * b->neuronCount);
			initializeNeuron(&(b->neurons[b->neuronCount-1]), b->neuronCount);
		}	
		else if(b->neuronCount > 1 + minInputCount +  minOutputCount) 
		{

			freeNeuron(&(b->neurons[b->neuronCount-1]));
			b->neuronCount--;
			b->neurons = realloc(b->neurons, sizeof(neuron)* b->neuronCount);

		}



	}

}


void targetMutation(brain * b, int targetCountLimit, float targetCountProbability, float targetCountBias , float retargetProbability){

	for(int i = 0 ; i < b->neuronCount; i++)
	{
		if(b->neurons[i].targetCount > 0)
		{
			if (targetCountProbability > randFloat())
			{




				if(targetCountBias > randFloat()  && b->neurons[i].targetCount < targetCountLimit)
				{
					b->neurons[i].targetCount++;
					b->neurons[i].targets = realloc(b->neurons[i].targets, sizeof(int) * b->neurons[i].targetCount);
					b->neurons[i].potentialWeights = realloc(b->neurons[i].potentialWeights,sizeof(float) * b->neurons[i].targetCount);
					b->neurons[i].potentialTimes = realloc(b->neurons[i].potentialTimes,sizeof(float) * b->neurons[i].targetCount);
					b->neurons[i].targets[b->neurons[i].targetCount-1] = randRange(b->neuronCount);
					b->neurons[i].potentialWeights[b->neurons[i].targetCount-1]  = randFloat();
				}
				else 	
				{

					b->neurons[i].targetCount--;
					b->neurons[i].targets = realloc(b->neurons[i].targets, sizeof(int) * b->neurons[i].targetCount);
					b->neurons[i].potentialWeights = realloc(b->neurons[i].potentialWeights,sizeof(float) * b->neurons[i].targetCount);
					b->neurons[i].potentialTimes = realloc(b->neurons[i].potentialTimes,sizeof(float) * b->neurons[i].targetCount);

				}



			}
			if(retargetProbability > randFloat() && b->neurons[i].targetCount>0)
			{
				b->neurons[i].targets[randRange(b->neurons[i].targetCount)] = randRange(b->neuronCount);
			}

		}
		else if(targetCountProbability > randFloat())
		{
			b->neurons[i].targetCount++;
			b->neurons[i].targets = realloc(b->neurons[i].targets, sizeof(int) * b->neurons[i].targetCount);
			b->neurons[i].potentialWeights = realloc(b->neurons[i].potentialWeights,sizeof(float) * b->neurons[i].targetCount);
			b->neurons[i].potentialTimes = realloc(b->neurons[i].potentialTimes,sizeof(float) * b->neurons[i].targetCount);
			b->neurons[i].targets[b->neurons[i].targetCount-1] = randRange(b->neuronCount);
			b->neurons[i].potentialWeights[b->neurons[i].targetCount-1]  = (randFloat()*2)-1;


		}
	}




}
/*
   Mutate Activation Potentials

probability: 0-1
strength: 0-infinity

 */


void potentialsMutation(brain * b, float probability, float strength){
	for(int i = 0 ; i < b->neuronCount; i++)
	{
		if(b->neurons[i].targetCount)
		{
			int mutations = randRange(b->neurons[i].targetCount);
			for (int c = 0 ; c < mutations; c++)
			{
				//TODO add soft boundary
				if(randFloat() < probability)
				{	
					b->neurons[i].potentialWeights[randRange(b->neurons[i].targetCount)] += ((randFloat() *2)-1)*strength;
				}
			}
		}
	}



}

void thresholdMutation(brain * b, float probability, float strength){
	for(int i = 0 ; i < b->neuronCount; i++)
	{
		if(b->neurons[i].targetCount)
		{
			int mutations = randRange(b->neurons[i].targetCount);
			for (int c = 0 ; c < mutations; c++)
			{
				//TODO add soft boundary
				if(randFloat() < probability)
				{	
					b->neurons[i].activationPotential += ((randFloat() *2)-1)*strength;
				}
			}
		}
	}



}




/*
   swaps some neurons


 */
void neuronSwapMutation(brain * b, float probability)
{
	if(randFloat() < probability)
	{
		int index1 = randRange(b->neuronCount-1)+1;	
		int index2 = randRange(b->neuronCount-1)+1;
		while(index2 == index1)
		{
			index2 = randRange(b->neuronCount-1)+1;
		}
		neuron temp;
		copyNeuron(&(b->neurons[index1]),&temp);
		copyNeuron(&(b->neurons[index2]),&(b->neurons[index1]));
		copyNeuron(&temp,&(b->neurons[index2]));




	}

}

/*
   randomly mutates a brain.
 */

void mutateBrain(brain * b, mutationParams * m){

	//historical presets, preserved for reference

	/*neuronCountMutation( b, minInputCount, minOutputCount, .125, .5);
	  targetMutation(b, .25, .5,.25);
	  potentialsMutation(b, 1, .1);
	  neuronSwapMutation(b, .1);*/



	/*
	   neuronSwapMutation(b, .1);
	   neuronCountMutation( b, minInputCount, minOutputCount, .125, .5);
	   targetMutation(b, .25, .5,.25);
	   potentialsMutation(b, 1, .1);
	   thresholdMutation(b,1,.1);
	 */

	char c;
	int i = 0;
	while((c = (m->order)[i]) != '\0'){
	i += 1;
	//printf("%c\n", c);
	switch (c) {


		case 's' :

			neuronSwapMutation(b,m->swapProb);
			break;

		case 'c' :

			neuronCountMutation(b, m->minInputCount, m->minOutputCount,m->neuronCountProb, m->neuronCountBias );
			break;

		case 't' : 

			targetMutation(b, m->targetLimit, m->targetCountProb, m->targetCountBias, m->retargetProb);
			break;

		case 'p' :

			potentialsMutation(b, m->potentialProb, m->potentialStrength);
			break;	
		case 'h' :

			thresholdMutation(b, m->potentialProb, m->potentialStrength);
			break;
	}



	}

}

/*
   applies inputs to a brain. inputs will be applied to neurons 1-targetCount

 */

void inputBrain(brain * b, int * inputs, int targetCount)
{

	assert(targetCount <= b->neuronCount-1);

	b->neurons[0].excitation = b->neurons[0].activationPotential + 1;// here we activate the bias neuron

	for(int i = 1; i < targetCount +1; i++) // the offsets here are to account for the bias node
	{

		b->neurons[i].excitation = inputs[i-1] * (b->neurons[i].activationPotential + 1) ;

	}

}

/*
   advances the brain forward 1 discrete fram of time

 */
void advanceBrain(brain * b, int  inputs[], int inputCount, int  outputs[], int outputCount)
{

	//if we've been passed a brain with too few neurons we will do nothing

	assert(! (b->neuronCount < inputCount+outputCount+1));


	inputBrain(b,inputs, inputCount);

	float * sums = malloc(sizeof(float) * b->neuronCount);
	for(int i = 0; i < b->neuronCount;i++)
	{
		sums[i] = 0.0;
	}

	for(int i = 0; i < b->neuronCount; i++)
	{
		b->neurons[i].fired = 0;
		if (mytanh(b->neurons[i].excitation) > b->neurons[i].activationPotential ){ // TODO: Think about the gate function (mytanh)
			for(int c = 0; c < b->neurons[i].targetCount; c++)
			{


				if(b->neurons[i].targets[c] < b->neuronCount)
				{
					sums[b->neurons[i].targets[c]] += b->neurons[i].potentialWeights[c];

				}
			}
			b->neurons[i].excitation = 0.0;	
			b->neurons[i].fired = 1;

		}

	}

	for(int i = 0; i < b->neuronCount; i++)
	{
		for(int c = 0; c < b->neurons[i].targetCount; c++)
		{

			if(b->neurons[i].targets[c] < b->neuronCount)
			{
				b->neurons[b->neurons[i].targets[c]].excitation += sums[b->neurons[i].targets[c]];
			}
		}

	}




	assert(outputCount + inputCount +1 <= b->neuronCount);
	int start = inputCount +1;
	for (int i = 0; i < outputCount; i++)
	{
		outputs[i] = b->neurons[start+i].fired;
	}
	free(sums);
}


/*
   displays a brain's structure and weights


 */
void printBrain(brain * b ){
	printf("ACTIVATION: ");
	for(int i = 0; i < b->neuronCount; i++)
	{
		printf("%d", b->neurons[i].fired);	
	}
	printf("\n");	
	for(int i = 0; i < b->neuronCount; i++)
	{
		printf("NEURON %d: %f\n",i, b->neurons[i].activationPotential);	
		for(int c = 0; c < b->neurons[i].targetCount; c++)
		{
			printf("TARGET: %d, WEIGHT :%f\n",b->neurons[i].targets[c],b->neurons[i].potentialWeights[c]);

		}	
	}



}

/*
   saves a brain to file

 */
void printBrainToFile(brain * b ,FILE * fp){
	fprintf(fp, "ACTIVATION: ");
	for(int i = 0; i < b->neuronCount; i++)
	{
		fprintf(fp,"%d", b->neurons[i].fired);	
	}
	printf("\n");	
	for(int i = 0; i < b->neuronCount; i++)
	{
		fprintf(fp, "NEURON %d: %f\n",i, b->neurons[i].activationPotential);	
		for(int c = 0; c < b->neurons[i].targetCount; c++)
		{
			fprintf(fp,"TARGET: %d, WEIGHT :%f\n",b->neurons[i].targets[c],b->neurons[i].potentialWeights[c]);

		}	
	}



}

File: python/brain.py
from random import uniform, randrange
from enum import Enum
import copy
import math
import utils
import json
import gym

class Brain_flags(Enum): ## dont change the existing ones without updating in mutation function
	NEURON_HIDDEN = 1
	NEURON_SENSORY = 2
	NEURON_ACTUATING = 3
	SENSOR_BIT = 4
	SENSOR_LINEAR_THRESHOLD = 5


class Neuron:
	
	def __init__(self):
		self.activation_potential = uniform(0,1)
		self.target_count = 0
		self.targets = []
		self.potential_weights = []
		self.fired = 0
		self.excitation = 0
		
		self.type = Brain_flags.NEURON_HIDDEN
		self.sensor_type = Brain_flags.SENSOR_BIT # not initalized to consistent vales(not used until neuron set to sensory or actuating)
		self.external_index = 0
		self.external_thresh = 0
		self.external_bit = 0



class Brain:

	def __init__(self, neuron_count):
		self.neuron_count = neuron_count
		self.neurons = []
		for i in range(neuron_count):			
			self.neurons.append(Neuron())
	
	def print_brain(self):
		print("ACTIVATION: ") 
		for neuron in self.neurons:
			print (str(neuron.fired)),
		print('')			
		for index, neuron in enumerate(self.neurons):
			print ("NEURON {index}".format(index = index))
			for i in range(neuron.target_count):
				print ("TARGET: {target}, WEIGHT: {weight}".format(target = neuron.targets[i],weight = neuron.potential_weights[i]))
		print('NEURON COUNT RECORD: ' +  str(self.neuron_count))
		print('NEURON COUNT: ' + str(len(self.neurons)))
			


	
	def sigmoid(self, x):
		 return 1 / (1 + math.exp(float( -x)))


	def hyperbolic_tangent(self, x):
		return 2*self.sigmoid(2*x)-1

	def verify_network_consistency(self):
		print("NEURON COUNT: " + str(self.neuron_count)) 

		print("NEURONS LENGTH: " + str(len(self.neurons))) 
		assert(self.neuron_count == len(self.neurons))
		for neuron in self.neurons:
			for target in neuron.targets:
				print('TARGET: ' + str(target))
				print("NEURON COUNT: " + str(self.neuron_count)) 
				assert(target < self.neuron_count)
		pass
		

	def neuron_count_mutation(self, min_input_count, min_output_count, probability, bias):

		if uniform(0,1) < probability:		
			if uniform(0,1) < bias:
				self.neurons.append(Neuron())	#needs neuron init function
				self.neuron_count  += 1
			elif self.neuron_count > 1:# + min_input_count + min_output_count:
				self.neuron_count -= 1
				del self.neurons[self.neuron_count]
				for neuron in self.neurons:
					for index, target in enumerate(neuron.targets):
						if target >= self.neuron_count:
							neuron.targets[index] = randrange(0,self.neuron_count)


	def target_mutation(self, max_targets, target_count_probability, target_count_bias, retarget_probability):
		if self.neuron_count > 1:
			for i, neuron in enumerate(self.neurons):
				if  target_count_probability > uniform(0,1):
					if target_count_bias > uniform(0,1) :
						if neuron.target_count < max_targets:
							neuron.target_count += 1
							new_target = randrange(0,self.neuron_count)
							while new_target == i:
								new_target = randrange(0,self.neuron_count)
							neuron.targets.append(new_target)
							neuron.potential_weights.append((uniform(0,1)*2)-1)
					else:
						if neuron.target_count > 0:
							neuron.target_count -= 1
							del neuron.targets[neuron.target_count]
							del neuron.potential_weights[neuron.target_count]


					if retarget_probability > uniform(0,1) and neuron.target_count > 0:
						neuron.targets[randrange(0,neuron.target_count)] = randrange(self.neuron_count)

	


	# note that the probabilities must add up to one or less here
	def type_mutation(self, input_count, output_count, sensory_mutation_prob, actuating_mutation_prob, hidden_mutation_prob):
		for neuron in self.neurons:
			selection = uniform(0,1)			
			if selection < sensory_mutation_prob:
				neuron.type = Brain_flags.NEURON_SENSORY
				neuron.sensor_type = Brain_flags(randrange(4,6))
				neuron.external_index = randrange(input_count)
				neuron.external_thresh = randrange(256)
				neuron.external_bit = randrange(8)
	
			elif selection < sensory_mutation_prob + actuating_mutation_prob:
				neuron.type = Brain_flags.NEURON_ACTUATING
				neuron.external_index = randrange(output_count)
		
		
			elif selection < sensory_mutation_prob + actuating_mutation_prob + hidden_mutation_prob:
				neuron.type = Brain_flags.NEURON_HIDDEN
			
			

					
	def potential_weights_mutation(self, probability, strength):

		for neuron in self.neurons:
			if neuron.target_count > 0:
				mutations = randrange(0, neuron.target_count)
				for i in range(mutations):
					if uniform(0,1) < probability:
						neuron.potential_weights[randrange(0,neuron.target_count)] += ((uniform(0,1) *2)-1)*strength

	def threshold_mutation(self, probability, strength):

		for neuron in self.neurons:
			if uniform(0,1) < probability:
				neuron.activation_potential += ((uniform(0,1) *2)-1)*strength


	def neuron_swap_mutation(self, probability):
		if self.neuron_count > 1:
			index1 = randrange(0, self.neuron_count)  
			index2 = randrange(0, self.neuron_count)  
			while index2 == index1:
				index2 = randrange(0, self.neuron_count) 
			temp = self.neurons[index1]
			self.neurons[index1] = self.neurons[index2]
			self.neurons[index2] = temp
			assert(len(self.neurons) == self.neuron_count)
			for index, neuron in enumerate(self.neurons):
				for i in range(neuron.target_count):
					while neuron.targets[i] == index:
						neuron.targets[i] = randrange(self.neuron_count) 

	
	def default_mutation(self, input_count, output_count):
		self.neuron_swap_mutation(.1)
		self.neuron_count_mutation(input_count, output_count, .125, .5)
		self.target_mutation(5,.25,.5,.25)
		self.potential_weights_mutation(1,.25)
		self.threshold_mutation(1,.5)
		self.type_mutation(input_count, output_count,.1,.1,.1)

	



	def input(self, inputs):
		self.neurons[0].excitation = self.neurons[0].activation_potential + 1  # activating the bias neuron



		for i in range(1, len(inputs) + 1 ):
			self.neurons[i].excitation = inputs[i-1] * (self.neurons[i].activation_potential + 1)
			
		
	def advance(self, inputs, output_length):

		#assert not (self.neuron_count < 1 + len(inputs) + output_length)
		#self.input(inputs)
		sums  = [0] * self.neuron_count
		outputs = [0] * output_length

		for i,neuron in enumerate(self.neurons):

			if neuron.fired:
				neuron.fired = 0
				if neuron.type == Brain_flags.NEURON_ACTUATING:
					outputs[neuron.external_index] = 1
	

			if neuron.type == Brain_flags.NEURON_SENSORY:
				observation = inputs[neuron.external_index]
				if neuron.sensor_type == Brain_flags.SENSOR_BIT:
					if (observation >> neuron.external_bit) & 1:	
						neuron.excitation = neuron.activation_potential + 1 
				elif self.neurons[i].sensor_type == Brain_flags.SENSOR_LINEAR_THRESHOLD:
					if observation > neuron.external_thresh:
						neuron.excitation = neuron.activation_potential + 1 


			elif self.hyperbolic_tangent(self.neurons[i].excitation) > self.neurons[i].activation_potential:
				for c in range (self.neurons[i].target_count):
					
					assert(self.neurons[i].targets[c] < self.neuron_count)
					sums[self.neurons[i].targets[c]] += self.neurons[i].potential_weights[c]
				self.neurons[i].fired = 1
				self.neurons[i].excitation = 0.0
			self.neurons[i].excitation = self.neurons[i].excitation/2.0
		
		for i in range(self.neuron_count):
			self.neurons[i].excitation += sums[i]
		#assert(len(inputs) + output_length + 1 <= self.neuron_count )
		#start = len(inputs) + 1
		#outputs = [0]
		#for i in range(output_length):
		#	outputs.append(self.neurons[start + i].fired)
		
		return outputs
	

def print_brain_to_file(brain):
	data = {}
	data['neuronCount'] = len(brain.neurons)
	data['neurons'] = []
	for neuron in brain.neurons:
		neuron_dict = {}
		neuron_dict['targets'] = neuron.targets
		neuron_dict['weights'] = neuron.potential_weights
		neuron_dict['thresh'] =  neuron.activation_potential
		neuron_dict['targetCount'] = neuron.target_count

		neuron_dict['neuronType'] = neuron.type.name
		neuron_dict['sensorType'] = neuron.sensor_type.name



		neuron_dict['externalIndex'] = neuron.external_index
		neuron_dict['externalThreshold'] = neuron.external_thresh
		neuron_dict['externalBit'] = neuron.external_bit
		data['neurons'].append(neuron_dict)

	output = json.dumps(data, indent = 4)
	output_file = open('log.txt', 'w')
	output_file.write(output)
	output_file.close()
	




#def load_brain_from_file():
	



